src/cr/cube/cube.py

```diff
@@ -806,4 +819,6 @@ class _WeightedCountMeasure(_BaseMeasure):
     @lazyproperty
     def _flat_values(self):
         """tuple of numeric counts after weighting."""
-        return tuple(self._cube_dict["result"]["measures"]["count"]["data"])
+        return np.array(
+            self._cube_dict["result"]["measures"]["count"]["data"], dtype=np.float64
+        )
```

Comment:
The docstring still says `tuple` but this is `1D np.float64 ndarray` now. But more importantly, I think this value needs to be optional now, like it is `None` if there is no `measures.count` field in the JSON.

Hmm, I suppose that's handled by the `_Measures.is_weighted` property that checks for this in the payload ... but that's awfully far away. Hmm, the broader structure of responsibilities is a little borked up it looks like, with like `_Measures.means` also checking the payload first.

I'd be inclined to think that `._flat_values` should be optional for _all_ measures that are sometimes not there, which I believe is all measures. Then if we want to know whether it has a measure we check whether that measure is None and we leave the payload interpretation to the actual measure class that can fully encapsulate it.

Not sure if you feel up to that refactoring just now, but I'd say at least that this `._flat_values` should be optional and maybe fix the spot in `_Measures.is_weighted` that inspects the payload for itself to rely on a value from `_WeightedCountMeasure` rather than poke around in the JSON. I suppose that means that `_BaseMeasure.raw_cube_array` would need to be optional too.

Hmm, this might require a little more noodling, and I don't want to launch you into a quagmire, but maybe you can have a think on it and we can spend a cafe minute on it or something. Either way, it can't hurt to start by making this optional.

Reply:
Okay, after reflection and nourishment, here's what I think:

1. All `._flat_values` are type `Optional 1D float64 ndarray` (maybe numeric array is 2D or something, I don't understand that enough to say, but ndarray anyway). The value is `None` when there is no such cube-measure in the payload or the values it needs are otherwise not to be found. This way, only the measure objects need to know where to go in the payload to get their values and all other measure-related payload inspection can be eliminated.

2. `_BaseMeasure.raw_cube_array` is also optional. Since this is the main interface method of a measure object (and `._flat_values` is private), it being `None` is how any higher-level unit figures out whether that base measure exists in the payload.

3. If we still need `_Measures.weighted_counts` to automatically substitute `_UnweightedCountMeasure` when the cube is unweighted, then we give `_Measures` a private `._weighted_counts` property which is either a `_WeightedCountMeasure` or `None` and let the interface property `.weighted_counts` perform the test and substitution, like `return self._weighted_counts if self._weighted_counts is not None else self.unweighted_counts`.

Overall, it still smells a little fishy, because I think that auto-substitution thing probably has to go, but this would at least remove part of the long-distance coupling and leakage of responsibilities.

Probably even better would be to add a `.load()` classmethod to `_BaseMeasure` and that is called for construction purposes rather than the subclass itself. That way it can return `None` if the measure cannot be created and callers don't have to test `.raw_cube_array` to see if their reference is valid or just an empty shell.

Let me know if this sounds like more than you bargained for and I can add a few-commit branch after yours merges. But it's not untypical that modules need some cleaning up before adding new features, so it wouldn't be a bad exercise to get this one tidied up a bit. This module mostly goes back to 2019 so it's a little bit crusty around the edges.

Reply:
Okay, I think I got this now